Quaver.Shared/Screens/Gameplay/Rulesets/Keys/HitObjects/GameplayHitObjectKeysInfo.cs

```diff
@@ -0,0 +1,179 @@
+
+        /// <summary>
+        ///     Latest position of this object.
+        ///
+        ///     For LNs with negative SVs, this can be larger than EndTrackPosition.
+        /// </summary>
+        public long LatestTrackPosition { get; private set; }
+
+        /// <summary>
+        ///     Earliest position of this object.
+        ///
+        ///     For LNs with negative SVs, this can be earlier than InitialTrackPosition.
+        /// </summary>
+        public long EarliestTrackPosition { get; private set; }
+
+        /// <summary>
+        ///     Constructor
+        /// </summary>
+        /// <param name="info"></param>
+        /// <param name="manager"></param>
+        /// <param name="state"></param>
+        /// <param name="hitObject"></param>
+		public GameplayHitObjectKeysInfo(HitObjectInfo info, HitObjectManagerKeys manager, HitObjectState state = HitObjectState.Alive, GameplayHitObjectKeys hitObject = null)
+		{
+            HitObjectInfo = info;
+            Manager = manager;
+			State = state;
+			HitObject = hitObject;
+
+			InitializePositions();
+		}
+
+        /// <summary>
+        ///     Calculates all relevant positions
+        /// </summary>
+		private void InitializePositions()
+		{
+			InitialTrackPosition = Manager.GetPositionFromTime(StartTime);
+
+            if (!IsLongNote)
+            {
+                LatestTrackPosition = InitialTrackPosition;
+            }
+            else
+            {
+                SVDirectionChanges = Manager.GetSVDirectionChanges(StartTime, EndTime);
+                EndTrackPosition = Manager.GetPositionFromTime(EndTime);
+
+                var earliestPosition = Math.Min(InitialTrackPosition, EndTrackPosition);
+                var latestPosition = Math.Max(InitialTrackPosition, EndTrackPosition);
+
+                foreach (var change in SVDirectionChanges)
+                {
+                    earliestPosition = Math.Min(earliestPosition, change.Position);
+                    latestPosition = Math.Max(latestPosition, change.Position);
+                }
+
+                EarliestTrackPosition = earliestPosition;
+                LatestTrackPosition = latestPosition;
+            }
+		}
+
+        /// <summary>
+        ///     Associate a GameplayHitObjectKeys to start drawing this hitobject.
+        /// </summary>
+        /// <param name="hitObject"></param>
+        public void Link(GameplayHitObjectKeys hitObject)
+        {
+			HitObject = hitObject;
+			HitObject.InitializeObject(Manager, this);
+        }
+
+        /// <summary>
+        ///     Dissaociate the currently linked GameplayHitObjectKeys to stop drawing this hitobject.
+        /// </summary>
+        /// <returns>The linked GameplayHitObjectKeys to add back to the pool.</returns>
+		public GameplayHitObjectKeys Unlink()
+		{
+			if (HitObject is null)
+				throw new InvalidOperationException("GameplayHitObjectInfo is not linked to a GameplayHitObjectKeys");
```

Comment:
In which case is this ever hit? You should handle this another way instead of crashing the game. Log and return perhaps? Not sure what the implications of that would be though.

Reply:
This case should only be hit if there's something else wrong with the code that assumes that the info is linked when it's unlinked. I figured throwing this exception would be more useful than getting a NullReferenceException. I guess this could be turned into a `bool TryUnlink(out GameplayHitObjectKeys)`, but I'd prefer the game just crash to make it obvious that there's a problem with the gameplay logic.

Reply:
Wrote a function to handle the logging and notification in the hit object manager.